src/storage/heap_file.c

```diff
@@ -24863,11 +24851,48 @@ heap_get_visible_version (THREAD_ENTRY * thread_p, const OID * oid, OID * class_
 */
 SCAN_CODE
 heap_scan_get_visible_version (THREAD_ENTRY * thread_p, const OID * oid, OID * class_oid, RECDES * recdes,
-			       HEAP_SCANCACHE * scan_cache, int ispeeking, int old_chn)
+			       RECDES * forward_recdes, HEAP_SCANCACHE * scan_cache, int ispeeking, int old_chn)
 {
   SCAN_CODE scan = S_SUCCESS;
   HEAP_GET_CONTEXT context;
 
+  /*
+   * Before fetching a record, check forward_recdes to see if the record type is REC_HOME,
+   * and it's being PEEKed (meaning there's no need to COPY the record data to a new space). 
+   * In this case, we can use forward_recdes as the record without executing the
+   * heap_get_visible_version_internal() function. If the conditions above are not met,
+   * or the mvcc_snapshot does not satisfy, then carry out the necessary steps through 
+   * the heap_get_visible_version_internal() function.
+   */
+  if (forward_recdes->type == REC_HOME && ispeeking == PEEK)
+    {
+      MVCC_REC_HEADER mvcc_header = MVCC_REC_HEADER_INITIALIZER;
+      if (or_mvcc_get_header (forward_recdes, &mvcc_header) != NO_ERROR)
+	{
+	  /* Unexpected. */
+	  assert (false);
+	}
+      if (scan_cache != NULL && scan_cache->mvcc_snapshot != NULL
+	  && scan_cache->mvcc_snapshot->snapshot_fnc != NULL && !mvcc_is_mvcc_disabled_class (class_oid)
+	  && scan_cache->mvcc_snapshot->snapshot_fnc (thread_p, &mvcc_header,
```

Comment:
I think that if you use 'scan_cache->mvcc_snapshot) == SNAPSHOT_SATISFIED' condition, then you can remove the redundant code.

Reply:
Nit:

1. assert (scan_cache != NULL); iso checking scan_cache != NULL; See the heap_next_internal().
2. assert (recdes != NULL); or checking it at the first if stmt.

Reply:
This change seems to change the logic. In the previous version, if `scan_cache == NULL`, `scan_cache->mvcc_snapshot == NULL`, `scan_cache->mvcc_snapshot->snapshot_fnc`, or `mvcc_is_mvcc_disabled_class`, it skips the visibility check. I thought that it was what you intended.  Now, it does the optimization only when the snapshot is satisfied. For example, this optimization worked for db_serial by `mvcc_is_mvcc_disabled_class()`, but it's not with the current version.

To me, it should be like
```
// as it was before and as in heap_get_visible_version_internal.
const need_check_visibility = scan_cache != NULL && scan_cache->mvcc_snapshot != NULL
 	  && scan_cache->mvcc_snapshot->snapshot_fnc != NULL && !mvcc_is_mvcc_disabled_class (class_oid)
 	  && scan_cache->mvcc_snapshot->snapshot_fnc (thread_p, &mvcc_header, scan_cache->mvcc_snapshot) != SNAPSHOT_SATISFIED; 

if (!need_check_visibility)
{
  *recdes = *forward_recdes;
  return scan;
}
/* fall through.. *//
} 

```